stdlib/Markdown/src/render/terminal/render.jl

```diff
@@ -16,118 +16,145 @@ end
 # TODO settle on another option with unicode e.g. "≡=≃–∼⋅" ?
 
-function term(io::IO, md::Header{l}, columns) where l
+function term(io::AnnotIO, md::Header{l}, columns) where l
+    face = Symbol("markdown_h$l")
     underline = _header_underlines[l]
-    _term_header(io, md, underline, columns)
+    pre = ' '^margin
+    local line_width
+    with_output_annotations(io, :face => face) do io
+        headline = annotprint(terminline, md.text)
+        lines = wraplines(headline, columns - 4margin)
+        for (i, line) in enumerate(lines)
+            print(io, pre, line)
+            i < length(lines) && println(io)
+        end
+        line_width = min(textwidth(lines[end]), columns)
+        if length(lines) > 1
+            line_width = max(textwidth(lines[end]), div(columns, 3)+length(pre))
+        end
+    end
+    header_width = max(0, line_width)
+    if underline != ' ' && header_width > 0
+        print(io, '\n', ' '^(margin))
+        with_output_annotations(io -> print(io, underline^header_width), io, :face => face)
+    end
 end
 
 function term(io::IO, md::Code, columns)
-    with_output_color(:cyan, io) do io
-        L = lines(md.code)
-        for i in eachindex(L)
-            print(io, ' '^margin, L[i])
-            i < lastindex(L) && println(io)
+    code = if md.language ∈ ("", "julia")
+        highlight(md.code)
+    elseif md.language == "julia-repl" || Base.startswith(md.language, "jldoctest")
+        hl = AnnotatedString(md.code)
+        for (; match) in eachmatch(r"(?:^|\n)julia>", hl)
+            StyledStrings.face!(match, :repl_prompt_julia)
+            afterprompt = match.offset + match.ncodeunits + 1
+            _, exprend = Meta.parse(md.code, afterprompt, raise = false)
+            highlight!(hl[afterprompt:prevind(md.code, exprend)])
+            if (nextspace = findnext(' ', md.code, exprend)) |> !isnothing
+                nextword = hl[exprend:prevind(hl, nextspace)]
+                if nextword == "ERROR:"
+                    StyledStrings.face!(nextword, :error)
+                end
+            end
         end
+        hl
+    else
+        styled"{markdown_code:$(md.code)}"
+    end
+    lines = split(code, '\n')
+    for (i, line) in enumerate(lines)
+        print(io, ' '^margin, line)
+        i < length(lines) && println(io)
     end
 end
 
 function term(io::IO, tex::LaTeX, columns)
-    printstyled(io, ' '^margin, tex.formula, color=:magenta)
+    print(io, ' '^margin, styled"{markdown_latex:$(tex.formula)}")
 end
 
 term(io::IO, br::LineBreak, columns) = nothing # line breaks already printed between subsequent elements
 
 function term(io::IO, br::HorizontalRule, columns)
-   print(io, ' '^margin, '─'^(columns - 2margin))
+    print(io, ' '^margin, styled"{markdown_hrule:$('─'^(columns - 2margin))}")
+end
+
+function term(io::IO, md::MarkdownElement, columns)
+    a = IOContext(AnnotatedIOBuffer(), io)
+    term(a, md, columns)
+    print(io, read(seekstart(a.io), AnnotatedString))
```

Comment:
Isn't there an optimized version of this, for copying between IO objects without the intermediate copy?
```suggestion
    write(io, seekstart(a.io))
```
Or is the problem that `io` might be wrapped up somewhat, so that dispatch currently won't reliably know if it is annotation-capable and end up discarding those?

Reply:
`io` may be `stdout`, and we *currently* don't have a specialised print/write method for `AnnotatedIOBuffer` to <some other buffer>.

I'd suggest that we perhaps add a "TODO change to direct write after implementing specialised AnnotatedIOBuffer printing"-type mental note or comment so we can come back to this if/when we do so.